monitor/base_monitor.go

```diff
@@ -0,0 +1,154 @@
+	warningAlertSent  bool
+}
+
+func (b *Base) Stop() {
+	b.RMC.StopChannel <- true
+}
+
+func (b *Base) Identify() string {
+	return b.Identifier
+}
+
+// Run the check on a given interval -> evaluate response via b.handle()
+func (b *Base) Run() error {
+	log.Debugf("%v-%v: Starting work for monitor %v...", b.Identify(), b.RMC.GID, b.RMC.Name)
+
+	defer b.RMC.Ticker.Stop()
+
+Mainloop:
+	for {
+		select {
+		case <-b.RMC.Ticker.C:
+			log.Debugf("%v-%v: Tick for monitor %v", b.Identify(), b.RMC.GID, b.RMC.Name)
+			if err := b.handle(b.MonitorFunc()); err != nil {
+				log.Errorf("Unable to complete check handler: %v", err.Error())
+			}
+		case <-b.RMC.StopChannel:
+			break Mainloop
+		}
+	}
+
+	log.Warningf("%v-%v: Goroutine has been stopped for monitor %v; exiting...", b.Identify(), b.RMC.GID, b.RMC.Name)
+	return nil
+}
+
+// Handle triggering/resolving alerts based on check results
+func (b *Base) handle(monitorErr error) error {
+	// Increase attempt count
+	b.attemptCount++
+
+	// No problems, reset counter
+	if monitorErr == nil {
+		// Send critical+warning resolve if critical threshold was exceed
+		// Send warning resolve if warning threshold was exceeded but critical threshold was not exceeded
+		if b.attemptCount > b.RMC.Config.CriticalThreshold {
+			alertMessage := fmt.Sprintf("Check has resolved after %v/%v attempts (critical)", b.attemptCount, b.RMC.Config.CriticalThreshold)
+			b.sendMessage(CRITICAL, alertMessage, true)
+			b.sendMessage(WARNING, alertMessage, true)
+		} else if b.attemptCount > b.RMC.Config.WarningThreshold {
+			alertMessage := fmt.Sprintf("Check has resolved after %v/%v attempts (warning)", b.attemptCount, b.RMC.Config.WarningThreshold)
+			b.sendMessage(WARNING, alertMessage, true)
+		}
+
+		b.attemptCount = 0
+
+		return nil
+	}
+
+	// Got an error; do we need to send any alerts?
+	if b.criticalAlertSent {
+		log.Debugf("Critical alert for %v already sent; skipping alerting", b.RMC.Name)
+		return nil
+	}
+
+	// Only return if we haven't reached critical threshold
+	if b.warningAlertSent && b.attemptCount < b.RMC.Config.CriticalThreshold {
+		log.Debugf("Warning alert for %v already sent; skipping alerting", b.RMC.Name)
+		return nil
+	}
+
+	// Okay, this must be the first time
+	if b.attemptCount == b.RMC.Config.CriticalThreshold {
+		alertMessage := fmt.Sprintf("Check has entered into critical state after %v checks (CriticalThreshold: %v)", b.attemptCount, b.RMC.Config.CriticalThreshold)
+		b.sendMessage(CRITICAL, alertMessage, false)
+	} else if b.attemptCount == b.RMC.Config.WarningThreshold {
+		alertMessage := fmt.Sprintf("Check has entered into warning state after %v checks (WarningThreshold: %v)", b.attemptCount, b.RMC.Config.WarningThreshold)
+		b.sendMessage(WARNING, alertMessage, false)
+	}
+
+	return nil
+}
+
+// Construct a new alert message, send down the message channel and update alert state
+func (b *Base) sendMessage(alertType int, alertMessage string, resolve bool) error {
```

Comment:
base method for sending alert messages - checks do not have to concern themselves about "how" to send an alert message, they just return an error.